lib/rules/no-loss-of-precision.js

```diff
@@ -0,0 +1,196 @@
+/**
+ * @fileoverview Rule to flag numbers that will lose significant figure precision at runtime
+ * @author Jacob Moore
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = {
+    meta: {
+        type: "problem",
+
+        docs: {
+            description: "disallow literal numbers that lose precision",
+            category: "Possible Errors",
+            recommended: true,
+            url: "https://eslint.org/docs/rules/no-loss-of-precision"
+        },
+        schema: [],
+        messages: {
+            noLossOfPrecision: "This number will lose precision when stored as a Number type."
+        }
+    },
+
+    create(context) {
+
+        /**
+         * Returns whether the node is number literal
+         * @param {Node} node the node literal being evaluated
+         * @returns {boolean} true if the node is a number literal
+         */
+        function isNumber(node) {
+            return typeof node.value === "number";
+        }
+
+        /**
+         * Checks whether the number is not base ten
+         * @param {ASTNode} node the node being evaluated
+         * @returns {boolean} true if the node is not in base ten
+         */
+        function isNotBaseTen(node) {
+            const prefixes = ["0x", "0X", "0b", "0B", "0o", "0O"];
+
+            return prefixes.some(prefix => node.raw.startsWith(prefix)) ||
+            (node.raw.startsWith("0") && !node.raw.startsWith("0."));
+        }
+
+        /**
+         * Checks that the user-intended non-base ten number equals the actual number after is has been converted to the Number type
+         * @param {Node} node the node being evaluated
+         * @returns {boolean} true if they do not match
+         */
+        function checkNotBaseTenNumber(node) {
+            const rawString = node.raw.toUpperCase();
+            let base = 0;
+
+            if (rawString.startsWith("0B")) {
+                base = 2;
+            } else if (rawString.startsWith("0X")) {
+                base = 16;
+            } else {
+                base = 8;
+            }
+
+            return !rawString.endsWith(node.value.toString(base).toUpperCase());
```

Comment:
Using `.endsWith` here seems a bit weird, although I can't think of any cases where it would be incorrect.

Can you add some tests that use leading zeros on non-base-10 literals (e.g. `0x005f`) to make sure the rule doesn't report an error?